### ISSUE TITLE
Question regarding the input dataset

### ISSUE THREAD

**Issue by pedromingues** [khirotaka/SAnD#7]
Hi,

I'm working on my thesis, and the aim is to study the progress of acute kidney injury up to several time steps ahead.

My plan is to use your code as the basis, and then add some things, like the ability to predict the progress of the disease while adding a level of uncertainty on every prediction, in order to know which predictions can be legit and trustworthy.

Having that, my first question is: 
Because not all itemids (features) have the same number of measurements, I believe you have to rearrange the events in time steps in order to fit in the input dimension you mentioned: **[N, seq_len, features]**. 
If so, did you produce time steps for every hour? Every 2 hours?

The second question is:
Did you run any specific benchmark to produce your dataset?

Best regards,
Pedro Domingues

**Comment by khirotaka** [comment]
Hi! Pedro Domingues-san

Thank you for your interest in my project.

> Because not all itemids (features) have the same number of measurements, I believe you have to rearrange the events in time steps in order to fit in the input dimension you mentioned: [N, seq_len, features].

Am I correct in understanding that "**the input data is information from multiple sensors, but each sensor has a different sampling frequency**"?

In this case, there are two ways I've come up with.

1. Use an interpolation function such as [`scipy.interpolate.interp1d()`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.interp1d.html) to align the lengths.
2. Prepare a SAnD model for each data of the same length, generate feature vectors, and concatenate them into a final representation.

Either way, you need to work on cutting out the time series using a sliding window. This is a kind of hyper-parameter, so you have to experiment with it.

> Did you run any specific benchmark to produce your dataset?

I don't remember if it was UCI/UEA or UCR, but I was experimenting for a while with the open ECG or Human Activity Recognition Dataset, which doesn't require registration and so on.

**Comment by khirotaka** [comment]
One week has passed. I decided that this issue has been resolved.

**Comment by pedromingues** [comment]
Hi, 

I'm still working on my thesis and I have a few specific (and maybe dumb) questions to ask..

During training, **does the model assume the whole batch of 128** (Using the batch size as in your example) **is a full sequence? Or does it consider each element of the batch as a sequence?** 
Still regarding that, **the Positional Encoding and the masked self-attention work on the whole batch of 128 elements, or do they 'occur' in each element of the batch?**

It might be a confusing question, but I ask this because I have my data in the format **[8732, 10, 36]**, which means **i have 8732 time stamps with patient data from their last 10 hours, taking into account 36 features**. I didn't have values for every feature in every hour, so I used interpolation functions to 'complete' my dataset.
**The problem is that those 8732 time stamps are not from the same patient. That number is from concatenating time stamps from several patients.** I have 30 time stamps from one patient and 12 from another, for example.

Should I be using different patients in the same batch during training? If the model assumes the whole batch is a sequence from the same patient it doesn't seem like a good idea. What do you recommend? Is my dataset in a wrong 'shape/format' to run on this model?

Also, you said that in your experiences:

> "SAnD is better at problems with a large number of features"

but do you still recommend a **feature selection** beforehand or having some features that probably wouldn't be important can actually help the model?

Just to end on a quick question, how do you want me to reference your work in my thesis? I thought about referencing this repository of yours but you might want something more specific :)

Thank you so much for your time and patience,
Pedro Domingues

**Comment by khirotaka** [comment]
Hi !!

Since SAnD has been proposed as a model for classification tasks, I will assume that your project is also a classification task.

> During training, does the model assume the whole batch of 128 (Using the batch size as in your example) is a full sequence? Or does it consider each element of the batch as a sequence?

The latter. The assumption is that each element of a batch is a single time series.
On the other hand, as in your example, there are some cases where the length of each sample varies and it is difficult to set the length to a fixed length. off course I have experimented with that case (I may not have published the code).

In this case, I created a partial time series using a sliding window and trained. It's worked fine.

So, for example
1. set up a window of length 24 (find the appropriate value through experimentation)
2. set the sliding width (or stride in CNN) (find the right width)
3. If the time series is shorter than the window size, you can fill it with zero or the previous value, etc. If the time series is larger than the window size, you can cut out multiple partial time series from a single sample.
4. train SAnD using the set of partial time series of equal length generated by 1~3. 

If I were in your position, I would try the above.

My experience with neural networks for time series classification (different from time series forecast, which predicts future values) is that when dealing with time series in a CNN or Transformer, it is best to think of them as 1D images.


> Still regarding that, the Positional Encoding and the masked self-attention work on the whole batch of 128 elements, or do they 'occur' in each element of the batch?

It applies to each element of the batch. This is because each element of the batch is supposed to be independent.

> but do you still recommend a feature selection beforehand or having some features that probably wouldn't be important can actually help the model?

I don't know if you should do a feature selection beforehand but if it's only 36 dimensions, I suppose you don't have to do it.
However, there is a way to make the feature selection explicit to the neural network: using [SE blocks](https://arxiv.org/abs/1709.01507). This is a method proposed for images, but it can be a pretty powerful tool for multidimensional time series classification as well. The SE block is implemented in [my public library](https://enchanter.readthedocs.io/en/stable/source/enchanter.addons.layers.html#selayer1d), so you can use it if you are interested :)

However, I have no idea where to put the SE block in SAnD. This will probably be a subject of research in itself.

> Just to end on a quick question, how do you want me to reference your work in my thesis? I thought about referencing this repository of yours but you might want something more specific :)

Please write my name(Hirotaka Kawashima) and the URL of this repository(https://github.com/khirotaka/SAnD) in the reference section of your paper. And if you can, I would be very happy if you could mention it in your paper with something like "In Hirotaka Kawashima's implementation" lol.

Sincerely.

Hirotaka Kawashima

**Comment by pedromingues** [comment]
Hi again ! 

I have another question, this one regarding the masked self-attention. 

Imagine I have a **sequence of 24 hours** and I'm trying to predict the following hour, with a mask value of 10. Does that mean **the model only looks at the last 10 hours to make the prediction**, ignoring the first 14 hours?

If yes, **are there any benefits using a sequence of 24 hours with a mask value of 10** comparatively to simply using a sequence of **10hours without any masking**?

Also, based on your personal tests and results, **did you find the model to work better with big or small sequences**?

Thank you once again for your time and patience.
Best regards,
Pedro Domingues

**Comment by khirotaka** [comment]
> Does that mean the model only looks at the last 10 hours to make the prediction, ignoring the first 14 hours?

There is a link in the code `subsequent_mask`, the Harvard material will give you a hint.

> Also, based on your personal tests and results, did you find the model to work better with big or small sequences?

Definitely a long time series data. As we already know from various studies, self-attention seems to be poor at capturing local features.
